### PR TITLE
Load state dict with assign to avoid OOMs

### LINKED ISSUE #480: [Question] load_state_dict: copy vs assign
### Question

Currently, as I understand, TransformerLens copies tensors when loading a pretrained model with
`HookedTransformer.from_pretrained `
by calling:
`load_and_process_state_dict -> self.load_state_dict(state_dict)`

Would it make sense to avoid copying and assign them instead? Latest Pytorch has a convenient option for this:

> https://pytorch.org/docs/2.1/generated/torch.nn.Module.html#torch.nn.Module.load_state_dict
> load_state_dict(state_dict, strict=True, assign=False)
> assign ([bool](https://docs.python.org/3/library/functions.html#bool), optional) – whether to assign items in the state dictionary to their corresponding keys in the module instead of copying them inplace into the module’s current parameters and buffers.

This would reduce memory use, but I'm not sure if there are any drawbacks -- what's the reason for copying?

Looks like the only issue is that `process_weights_` feature would not work, which might be fine, optionally?

### PR DESCRIPTION
**Pull request by cyber-chris**
# Description

I personally keep getting OOMs when trying to load llama3 8B. I narrowed down the offending code to `load_and_process_state_dict`, then realized Neel offered a suggestion in #480 that isn't implemented yet.

So, a simple change: always use `assign=True` (if the installed PyTorch version is recent enough).

Previously, tests broke because of a mismatch in the `dtype` of the state_dict mask and the default TL mask. Adding a cast fixes all tests.

Fixes #480. (issue)

## Type of change

Bug fix or new feature, depending on perspective. Should allow loading more models without needing a ton of memory.

### Screenshots
Please attach before and after screenshots of the change if applicable.

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

File: transformer_lens/components/abstract_attention.py [hunk 1]
```diff
@@ -451,7 +451,9 @@ def apply_causal_mask(
             )
 
         # Index back to front to ensure local attention works
-        final_mask = self.mask[None, None, -query_ctx_length:, -key_ctx_length:]  # [1, 1, pos, pos]
+        final_mask = self.mask[
+            None, None, -query_ctx_length:, -key_ctx_length:
+        ].bool()  # [1, 1, pos, pos]
```
Comment (cyber-chris) [inline transformer_lens/components/abstract_attention.py:456]: The `dtype` of the mask loaded from the state dict is float, but this class expects it to be a bool tensor. Casting to bool just before applying the mask fixes all the tests.
image

**Comment by bryce13950** [comment]
Thank you very much for doing this! I am going to be using this PR to work on something else I have been working on to see if I notice anything odd. This is a pretty sensitive area of the project, so we need to make sure things are working as expected before merging.

**Comment by bryce13950** [comment]
Alright, so after first run, it seems pretty good. I did get some weird generations with your branch though, which is generally not unheard of. I do want to run my two scripts side by side to make sure there is no further degradation beyond what we have at the moment.

**Comment by bryce13950** [comment]
Well, I am sorry to say, but there are more issues. Both `Qwen/Qwen-1_8B` and `Qwen/Qwen-1_8B-Chat` are both broken on this branch. I have verified that they both work with the current release of TransformerLens, and that the failure is indeed exclusive to this PR. The issue is type mismatch.

Here is the full error
```
[/usr/local/lib/python3.10/dist-packages/transformer_lens/HookedTransformer.py](https://8toay7rrexo-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab_20240919-060125_RC00_676352627#) in generate(self, input, max_new_tokens, stop_at_eos, eos_token_id, do_sample, top_k, top_p, temperature, freq_penalty, use_past_kv_cache, prepend_bos, padding_side, return_type, verbose)
   2137                         )
   2138                     else:
-> 2139                         logits = self.forward(
   2140                             tokens,
   2141                             return_type="logits",

[/usr/local/lib/python3.10/dist-packages/transformer_lens/HookedTransformer.py](https://8toay7rrexo-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab_20240919-060125_RC00_676352627#) in forward(self, input, return_type, loss_per_token, prepend_bos, padding_side, start_at_layer, tokens, shortformer_pos_embed, attention_mask, stop_at_layer, past_kv_cache)
    559                     )
    560 
--> 561                 residual = block(
    562                     residual,
    563                     # Cache contains a list of HookedTransformerKeyValueCache objects, one for each

[/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py](https://8toay7rrexo-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab_20240919-060125_RC00_676352627#) in _wrapped_call_impl(self, *args, **kwargs)
   1551             return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1552         else:
-> 1553             return self._call_impl(*args, **kwargs)
   1554 
   1555     def _call_impl(self, *args, **kwargs):

[/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py](https://8toay7rrexo-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab_20240919-060125_RC00_676352627#) in _call_impl(self, *args, **kwargs)
   1560                 or _global_backward_pre_hooks or _global_backward_hooks
   1561                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1562             return forward_call(*args, **kwargs)
   1563 
   1564         try:

[/usr/local/lib/python3.10/dist-packages/transformer_lens/components/transformer_block.py](https://8toay7rrexo-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab_20240919-060125_RC00_676352627#) in forward(self, resid_pre, shortformer_pos_embed, past_kv_cache_entry, attention_mask)
    158             # queries, keys and values, independently.
    159             # Then take the layer norm of these inputs, and pass these to the attention module.
--> 160             self.attn(
    161                 query_input=self.ln1(query_input)
    162                 + (0.0 if shortformer_pos_embed is None else shortformer_pos_embed),

[/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py](https://8toay7rrexo-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab_20240919-060125_RC00_676352627#) in _wrapped_call_impl(self, *args, **kwargs)
   1551             return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1552         else:
-> 1553             return self._call_impl(*args, **kwargs)
   1554 
   1555     def _call_impl(self, *args, **kwargs):

[/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py](https://8toay7rrexo-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab_20240919-060125_RC00_676352627#) in _call_impl(self, *args, **kwargs)
   1560                 or _global_backward_pre_hooks or _global_backward_hooks
   1561                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1562             return forward_call(*args, **kwargs)
   1563 
   1564         try:

[/usr/local/lib/python3.10/dist-packages/transformer_lens/components/abstract_attention.py](https://8toay7rrexo-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab_20240919-060125_RC00_676352627#) in forward(self, query_input, key_input, value_input, past_kv_cache_entry, additive_attention_mask, attention_mask, position_bias)
    193         """
    194 
--> 195         q, k, v = self.calculate_qkv_matrices(query_input, key_input, value_input)
    196 
    197         if past_kv_cache_entry is not None:

[/usr/local/lib/python3.10/dist-packages/transformer_lens/components/abstract_attention.py](https://8toay7rrexo-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab_20240919-060125_RC00_676352627#) in calculate_qkv_matrices(self, query_input, key_input, value_input)
    354             )
    355         else:
--> 356             q = self.hook_q(attn_fn(query_input, self.W_Q, self.b_Q))
    357         if self.cfg.load_in_4bit:
    358             if not isinstance(self.W_K, Params4bit):

[/usr/local/lib/python3.10/dist-packages/transformer_lens/utilities/attention.py](https://8toay7rrexo-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab_20240919-060125_RC00_676352627#) in simple_attn_linear(input, w, b)
     17     w = einops.rearrange(w, "head_index d_model d_head -> (head_index d_head) d_model")
     18     b_ = einops.rearrange(b, "head_index d_head -> (head_index d_head)")
---> 19     return F.linear(input, w, b_).reshape(input.shape[0], input.shape[1], b.shape[0], b.shape[1])
     20 
     21 

RuntimeError: mat1 and mat2 must have the same dtype, but got Float and BFloat16
```

To be completely honest with you, I am not sure if this is worth pursuing. There could be a whole cascade of issues here caused by this loading variable, which we may not even think to test due to the multitudes of possible ways TransformerLens can be configured. I am not sure if I would ever feel comfortable merging this given the whack-a-mole bug hunting that this is already turning into. The CI coverage is simply not high enough for me to feel like we will ever have all of the bugs caused by this resolved. 

If you want to keep pursuing this, we can, and I am happy to keep stress testing different revisions to make sure it is working, but I want you to know my feeling today before you spend another 10-20 hours working on this to ultimately have it be decided to not merge 2 months from now due to compatibility uncertainty. 

Realistically, this [pr](https://github.com/TransformerLensOrg/TransformerLens/pull/706) got us 98% of the way there. Using this variable is going to use slightly less ram than what TransformerLens uses now, but only slightly less. Three months ago, this would have saved 50% memory consumption, but things have changed since then. This variable will cause the load to run faster, but when weighing a change on the spectrum of compatibility/stability against performance, compatibility is always going to win out on this project. 

If we can improve performance even 1% without impacting compatibility, then that's great. The moment we start seeing that something is a miss in the compatibility, then that raises a massive red flag. Now we are in a situation where the CI is passing, but we don't have great coverage to begin with, and it was really easy for me to break this implementation beyond what the CI tells us. 

It's a good idea, and a good effort, but I am very uncertain of if the benefits of this outweighs the potential pile of bugs we may introduce.

**Comment by cyber-chris** [comment]
Thanks, closing this for now. I was pondering if there is some way to only use `assign` if the `dtype` matches what's expected by TL, but not sure how to properly do that kind of type introspection.

**Comment by bryce13950** [comment]
@cyber-chris If you want to explore this more later on let me know! I may still reopen your branch, and merge it into the main repo as an experiment. I am working on a way to bulk test all supported models, which is a pretty big logistical task. Once that is done, I would be more willing to put something like this into a release.

**Pull request closed without merging**